File: src/main/java/com/github/nsc/de/shake/lexer/characterinputstream/position/PositionMap.java
package com.github.nsc.de.shake.lexer.characterinputstream.position;

import com.github.nsc.de.shake.lexer.characterinputstream.charactersource.CharacterSource;

public class PositionMap {

    private final CharacterSource source;
    private final int[] lineSeparators;


    public PositionMap(CharacterSource source, int[] lineSeparators) {
        this.source = source;
        this.lineSeparators = lineSeparators;
    }

    public CharacterSource getSource() {
        return source;
    }

    public String getLocation() {
        return getSource().getLocation();
    }

    public int[] getLineSeparators() {
        return lineSeparators;
    }

    public Position resolve(int index) {

        for(int i = 0; i < this.getLineSeparators().length; i++) {
            if(index < this.getLineSeparators()[i]) {
                if(i == 0) return new Position(this, index, index + 1, 1);
                else return new Position(this, index, index - this.getLineSeparators()[i - 1], i + 1);
            }
        }

        return new Position(this, index,
                index - (this.getLineSeparators().length > 0 ? this.getLineSeparators()[this.getLineSeparators().length - 1] : 0) + 1,
                this.getLineSeparators().length + 1);

    }

    public int getAfterInLine(Position p) {

        return p.getLine() - 1 == lineSeparators.length ?
                getSource().getLength() - p.getColumn() :
                lineSeparators[p.getLine() - 1] - p.getIndex();

    }

    @Override
    public String toString() {
        return getLocation();
    }
}

File: util/parseutils/src/commonMain/kotlin/io/github/shakelang/shake/util/parseutils/characters/position/PositionMap.kt
package io.github.shakelang.shake.util.parseutils.characters.position

import io.github.shakelang.shake.util.parseutils.characters.source.CharacterSource

interface PositionMap {
    val source: CharacterSource
    val lineSeparators: IntArray
    val location: String get() = source.location

    fun resolve(index: Int): Position {
        for (i in lineSeparators.indices) {
            if (index < lineSeparators[i] + 1) {
                return if (i == 0) Position(
                    this,
                    index,
                    index + 1,
                    1
                ) else Position(
                    this,
                    index,
                    index - lineSeparators[i - 1],
                    i + 1
                )
            }
        }
        return Position(
            this, index,
            index - (if (lineSeparators.isNotEmpty()) lineSeparators[lineSeparators.size - 1] + 1 else 0) + 1,
            lineSeparators.size + 1
        )
    }

    fun getAfterInLine(p: Position): Int {
        return if (p.line - 1 == lineSeparators.size) source.length - p.column else lineSeparators[p.line - 1] - p.index
    }
    fun getAfterInLine(index: Int): Int = getAfterInLine(resolve(index))

    open class PositionMapImpl(
        override val source: CharacterSource,
        override val lineSeparators: IntArray
    ) : PositionMap {

        override val location: String
            get() = source.location

        override fun getAfterInLine(p: Position): Int {
            return if (p.line - 1 == lineSeparators.size) source.length - p.column else lineSeparators[p.line - 1] - p.index
        }

        override fun toString(): String {
            return location
        }
    }
}